blockchain/chain_sync/src/validation.rs

```diff
@@ -0,0 +1,137 @@
+        Self::IpldAmt(err.to_string())
+    }
+}
+
+impl From<EncodingError> for TipsetValidationError {
+    fn from(err: EncodingError) -> Self {
+        Self::Encoding(err)
+    }
+}
+
+pub struct TipsetValidator<'a>(pub &'a FullTipset);
+
+impl<'a> TipsetValidator<'a> {
+    pub async fn validate<DB: BlockStore + Send + Sync + 'static>(
+        &self,
+        chainstore: Arc<ChainStore<DB>>,
+        bad_block_cache: Arc<BadBlockCache>,
+        genesis_tipset: Arc<Tipset>,
+    ) -> Result<(), TipsetValidationError> {
+        // No empty blocks
+        if self.0.blocks().is_empty() {
+            return Err(TipsetValidationError::NoBlocks);
+        }
+
+        // Tipset epoch must not be behind current max
+        self.validate_epoch(genesis_tipset)?;
+
+        // Validate each block in the tipset by:
+        // 1. Calculating the message root using all of the messages to ensure it matches the mst root in the block header
+        // 2. Ensuring it has not previously been seen in the bad blocks cache
+        for block in self.0.blocks() {
+            self.validate_msg_root(chainstore.db.as_ref(), block)?;
+            if let Some(bad) = bad_block_cache.peek(block.cid()).await {
+                return Err(TipsetValidationError::InvalidBlock);
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn validate_epoch(&self, genesis_tipset: Arc<Tipset>) -> Result<(), TipsetValidationError> {
+        let now = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_secs();
+        let max_epoch =
+            ((now - genesis_tipset.min_timestamp()) / BLOCK_DELAY_SECS) + MAX_HEIGHT_DRIFT;
+        let too_far_ahead_in_time = self.0.epoch() as u64 > max_epoch;
+        if too_far_ahead_in_time {
+            return Err(TipsetValidationError::EpochTooLarge);
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn validate_msg_root<DB: BlockStore>(
+        &self,
+        blockstore: &DB,
+        block: &Block,
+    ) -> Result<(), TipsetValidationError> {
+        let msg_root = Self::compute_msg_root(blockstore, block.bls_msgs(), block.secp_msgs())?;
+        if block.header().messages() != &msg_root {
+            Err(TipsetValidationError::InvalidRoots)
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn compute_msg_root<DB: BlockStore>(
+        blockstore: &DB,
+        bls_msgs: &[UnsignedMessage],
+        secp_msgs: &[SignedMessage],
+    ) -> Result<Cid, TipsetValidationError> {
+        // Generate message CIDs
+        let bls_cids = bls_msgs
+            .iter()
+            .map(Cbor::cid)
+            .collect::<Result<Vec<Cid>, EncodingError>>()?;
+        let secp_cids = secp_msgs
+            .iter()
+            .map(Cbor::cid)
+            .collect::<Result<Vec<Cid>, EncodingError>>()?;
+
+        // Generate Amt and batch set message values
+        let bls_message_root = Amt::new_from_iter(blockstore, bls_cids)?;
```

Comment:
Is it possible to create Amt with iterators of bls_cids. secp_cids as the method suggests so? We are collecting them as a Vec here

Reply:
Is the suggestion to call `.iter()` on the vectors are they are being passed as arguments in the function?

Reply:
no, without collecting it and just mapping:
```rust
Amt::new_from_iter(bls_msgs.iter().map(Cbor::cid))
```

Reply:
I think the problem with that is that the map produces a `Result<Cid,_>` instead of a `Cid` and the `Amt::new_from_iter` method accepts an `IntoIterator<Item = V>`